.emacs.d/config.org

```diff
@@ -55,62 +55,392 @@
    #+end_src
+
+* Projects
+
+  - https://github.com/bbatsov/projectile
+  - https://www.projectile.mx/en/latest/
+
+  Projectile gives Emacs project management features, such as navigating files, finding occurences of a string, running build commands and more.
+  It detects project folders through version control (git, svn, ...) and you can optionally give a default project folder path within the config.
+
+  #+begin_src emacs-lisp
+  (use-package projectile
+    ;; Getting complaints about undefined variables when these are placed in
+    ;; `:custom`.
+    :config
+    ;; Useful for any CMake-based project, and using `ccls` with C/C++
+    (setq projectile-project-root-files-top-down-recurring
+     (append
+      '("compile_commands.json" ".ccls")
+      projectile-project-root-files-top-down-recurring))
+    (setq projectile-globally-ignored-directories
+     (append
+      '("build" "CMakeFiles" ".ccls-cache")
+      projectile-globally-ignored-directories))
+    (setq projectile-globally-ignored-files
+     (append
+      '("cmake_install.cmake")
+      projectile-globally-ignored-files))
+    :custom
+    (projectile-project-search-path '("~/Projects/"))
+    (projectile-enable-caching t)
+    (projectile-require-project-root t)
+    (projectile-sort-order 'access-time)
+    :init
+    (projectile-mode))
+  #+end_src
+
+* Magit
+
+  [[https://github.com/magit/magit][Magit]] is a Git interface, just like you have in many IDEs and text editors, but combining that with all the power of using Git from the command line.
+
+  #+begin_src emacs-lisp
+  (use-package magit)
+  #+end_src
+
+* Key bindings
+
+  [[https://github.com/justbur/emacs-which-key][Which-key]] is a package that displays what bindings are available when you start pressing a key. It is incredibly useful, especially when you are new to Emacs or when you are trying a new package.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package which-key
+    :init
+    (which-key-mode))
+  #+END_SRC
+
+* Linters
+
+  The two main frameworks that allow Emacs to interface with external linters are
+  - [[https://www.gnu.org/software/emacs/manual/html_node/emacs/Flymake.html][flymake]]
+  - [[https://www.flycheck.org/en/latest/][flycheck]]
+  A detailed comparison is at https://www.flycheck.org/en/latest/user/flycheck-versus-flymake.html. We choose to use flycheck due to the sheer number of extensions written for it; pick your favorite language linter, and there’s probably a flycheck extension for it.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package flycheck)
+  #+END_SRC
+
+* Language and debug servers
+
+  [[https://microsoft.github.io/language-server-protocol][LSP]] is a protocol used to implement IDE-like features in an editor-agnostic manner. This package will give us a uniform way to add features such as semantic highlighting, auto-completion, jump-to-definition, and so on.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-mode
+    :commands lsp
+    :hook ((c-mode c++-mode python-mode sh-mode) . lsp)
```

Comment:
You don't need this as we have hooks defined per language later on..

Reply:
The reason hooks are defined per-language later on is because those are for non-core language server configurations that don't provide autoloads. Let's say you wanted to use `ccls` instead of `clangd`. You would remove `c-mode c++-mode` from here and add them to the `ccls` hooks, unless you wanted both language servers to be selectable with one chosen over the other based on the priority value (like in https://github.com/emacs-lsp/lsp-python-ms/blob/83ef84d9a4a942f8882b00d07bb78b15f716e89d/lsp-python-ms.el#L432 vs https://github.com/emacs-lsp/lsp-mode/blob/1e49cedd903727d511a8e90ada9ce9d3f84f6b80/lsp-pyls.el#L408).

The only time you wouldn't need a hook at all is if the LSP package (either core or external) autoloaded a hook for the respective language mode, which none of them do.

Reply:
I think I was misunderstood. I understand the ccls lambda hooks due to our private conversation, but in order to use say clangd, I was pointing out it's cleaner to do:
```cpp
(use-package cc-mode
  :ensure nil
  :hook ((c-mode c++-mode) . lsp))
```